bofire/strategies/stepwise/stepwise.py

```diff
@@ -0,0 +1,75 @@
+from typing import Dict, Tuple, Type
+
+import pandas as pd
+
+import bofire.data_models.strategies.api as data_models
+import bofire.strategies.stepwise.conditions as conditions
+from bofire.data_models.strategies.api import Step
+from bofire.data_models.strategies.api import StepwiseStrategy as data_model
+from bofire.strategies.doe_strategy import DoEStrategy  # noqa: F401
+from bofire.strategies.predictives.qehvi import QehviStrategy  # noqa: F401
+from bofire.strategies.predictives.qnehvi import QnehviStrategy  # noqa: F401
+from bofire.strategies.predictives.qparego import QparegoStrategy  # noqa: F401
+from bofire.strategies.predictives.sobo import (  # noqa: F401
+    AdditiveSoboStrategy,
+    MultiplicativeSoboStrategy,
+    SoboStrategy,
+)
+from bofire.strategies.random import RandomStrategy  # noqa: F401
+from bofire.strategies.samplers.polytope import PolytopeSampler  # noqa: F401
+from bofire.strategies.samplers.rejection import RejectionSampler  # noqa: F401
+from bofire.strategies.strategy import Strategy  # noqa: F401
+
+# we have to duplicate the map functionality due to prevent circular imports
+STRATEGY_MAP: Dict[Type[data_models.Strategy], Type[Strategy]] = {
+    data_models.RandomStrategy: RandomStrategy,
+    data_models.SoboStrategy: SoboStrategy,
+    data_models.AdditiveSoboStrategy: AdditiveSoboStrategy,
+    data_models.MultiplicativeSoboStrategy: MultiplicativeSoboStrategy,
+    data_models.QehviStrategy: QehviStrategy,
+    data_models.QnehviStrategy: QnehviStrategy,
+    data_models.QparegoStrategy: QparegoStrategy,
+    data_models.PolytopeSampler: PolytopeSampler,
+    data_models.RejectionSampler: RejectionSampler,
+    data_models.DoEStrategy: DoEStrategy,
+}
+
+
+def map(data_model: data_models.Strategy) -> Strategy:
+    cls = STRATEGY_MAP[data_model.__class__]
+    return cls.from_spec(data_model=data_model)
+
+
+class StepwiseStrategy(Strategy):
+    def __init__(self, data_model: data_model, **kwargs):
+        super().__init__(data_model, **kwargs)
+        self.steps = data_model.steps
+
+    def has_sufficient_experiments(self) -> bool:
+        return True
+
+    def _get_step(self) -> Tuple[int, Step]:  # type: ignore
+        for i in range(len(self.steps) - 1, -1, -1):
```

Comment:
Why do you start with the last step instead of the first step?

Reply:
For me it is easier like this:  the step that should be applied is always the step which comes before the first step with an unfullfilled condition. 

For example, if you have three steps and the conditions looks like this `[True, True, False]`, then we would proceed with the step with index 10`. For me it is then easier to loop from the back and stop with the first condition where one gets a fulfilled condition instead starting from the front and looking for the first `False` and then returning the step before the `False`.

Do you agree?

Reply:
I would start with the first step that evaluates to `True` and care about the later steps when it is there turn. This seems simpler to me. When the condition of the first step changes to `False`  I would go to the next step. Basically, as in your initial version with the only difference that numbers of experiments is not the only possible criterion.

Reply:
Hmm, but this would change the whole logic of the implemented conditions. For example the `RequiredExperimentsCondition` with `n_required_experiments` will neve turn to `False` when it is once turned `True`. For this reason, I start from the back to always use the latest applicable step.

Reply:
Ah, yes. That's right. I vote for changing the inequality-sign in `RequiredExperimentsCondition` and I would rename it to `NumberOfExperimentsCondition`. The reason why I think we should start with the first step is that it is what people are used to. Starting with the first step. I think it will be confusing to start with the last step.

Reply:
Ok, I will change it.

Reply:
I implemented it according to your suggestion, should we then for the `NumberOfExperimentsCondition` implement the possibility of infinity experiments, so that it holds true forever, so that the user can decide if he wants to end an campaign with a step or not ...

one could encode it via `n_experiments = -1`

What do you think?

Reply:
Ah. You mean the last step could be forever? Yes, we can do that. For this an always `True` condition might be more explicit. But your suggestion is also fine.

Reply:
Yes, last step forever ;) I like the idea of an `AlwaysTrueCondition`. I will add it, and validation that this can only be used in the last step.

Reply:
@bertiqwerty: Done, now it should be finally ready. Can you have a look again?